src/server/routes/setPassword.ts

```diff
@@ -0,0 +1,150 @@
+import { Request, Router } from 'express';
+import { Routes } from '@/shared/model/Routes';
+import { renderer } from '@/server/lib/renderer';
+import { ResponseWithRequestState } from '@/server/models/Express';
+import { PageTitle } from '@/shared/model/PageTitle';
+import {
+  checkResetPasswordTokenController,
+  setPasswordTokenController,
+} from '@/server/controllers/changePassword';
+import { handleAsyncErrors } from '@/server/lib/expressWrappers';
+import deepmerge from 'deepmerge';
+import { RequestError } from '@/shared/lib/error';
+import { logger } from '@/server/lib/logger';
+import { setEncryptedStateCookie } from '@/server/lib/encryptedStateCookie';
+import { EmailType } from '@/shared/model/EmailType';
+import { sendCreatePasswordEmail } from '@/server/lib/idapi/user';
+import { readEmailCookie } from '@/server/lib/emailCookie';
+import { addReturnUrlToPath } from '@/server/lib/queryParams';
+
+const router = Router();
+
+// set password complete page
+router.get(
+  Routes.SET_PASSWORD_COMPLETE,
+  (req: Request, res: ResponseWithRequestState) => {
+    const email = readEmailCookie(req);
+
+    const html = renderer(Routes.SET_PASSWORD_COMPLETE, {
+      requestState: deepmerge(res.locals, {
+        pageData: {
+          email,
+        },
+      }),
+      pageTitle: PageTitle.SET_PASSWORD_COMPLETE,
+    });
+    return res.type('html').send(html);
+  },
+);
+
+// resend "create (set) password" email page
+router.get(
+  `${Routes.SET_PASSWORD}${Routes.RESEND}`,
+  (_: Request, res: ResponseWithRequestState) => {
+    const html = renderer(`${Routes.SET_PASSWORD}${Routes.RESEND}`, {
+      pageTitle: PageTitle.SET_PASSWORD_RESEND,
+      requestState: res.locals,
+    });
+    res.type('html').send(html);
+  },
+);
+
+// set password page session expired
+router.get(
+  `${Routes.SET_PASSWORD}${Routes.EXPIRED}`,
+  (_: Request, res: ResponseWithRequestState) => {
+    const html = renderer(`${Routes.SET_PASSWORD}${Routes.EXPIRED}`, {
+      pageTitle: PageTitle.SET_PASSWORD_RESEND,
+      requestState: res.locals,
+    });
+    res.type('html').send(html);
+  },
+);
+
+// POST handler for resending "create (set) password" email
+router.post(
+  `${Routes.SET_PASSWORD}${Routes.RESEND}`,
+  handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+    const { email } = req.body;
+    const { returnUrl } = res.locals.queryParams;
+
+    try {
+      await sendCreatePasswordEmail(email, req.ip, returnUrl);
+
+      setEncryptedStateCookie(res, {
+        email,
+        emailType: EmailType.CREATE_PASSWORD,
+      });
+
+      return res.redirect(303, `${Routes.SET_PASSWORD_EMAIL_SENT}`);
+    } catch (error) {
+      const { message, status } = error as RequestError;
```

Comment:
what if the error is not a RequestError?
If a standard JS error is caught here, then it could have a message with status undefined

Reply:
added some defaults in to return to the user, logs will show true error if needed

Reply:
We should try to avoid surfacing technical error messages to users. Perhaps the handled api errors should use a JavaScript class instead of a `type`, so we can check the error class name at runtime.

Reply:
We're not surfacing any technical error messages to users, just a generic "We had trouble resetting your password". The `logger `method will log the actual error to kibana if we need to read it.

I do agree though that it would be useful to refactor the errors in the long term, especially to differentiate between javascript runtime errors, and api errors from idapi/okta in the future.

Reply:
Note: Errors were refactored in https://github.com/guardian/gateway/pull/1104